lib/instrumentation/modules/kafkajs.js

```diff
@@ -0,0 +1,451 @@
+    return async function (payload) {
+      const { topic, message } = payload;
+
+      if (shouldIgnoreTopic(topic, config)) {
+        return origEachMessage.apply(this, arguments);
+      }
+
+      const traceparent = message.headers && message.headers.traceparent;
+      const tracestate = message.headers && message.headers.tracestate;
+      const opts = {};
+
+      // According to `kafkajs` types a header value might be
+      // a string or Buffer
+      // https://github.com/tulios/kafkajs/blob/ff3b1117f316d527ae170b550bc0f772614338e9/types/index.d.ts#L148
+      if (typeof traceparent === 'string') {
+        opts.childOf = traceparent;
+      } else if (traceparent instanceof Buffer) {
+        opts.childOf = traceparent.toString('utf-8');
+      }
+
+      if (typeof tracestate === 'string') {
+        opts.tracestate = tracestate;
+      } else if (tracestate instanceof Buffer) {
+        opts.tracestate = tracestate.toString('utf-8');
+      }
+
+      const trans = ins.startTransaction(
+        `${NAME} RECEIVE from ${topic}`,
+        TYPE,
+        opts,
+      );
+
+      const messageCtx = { queue: { name: topic } };
+      if (
+        config.captureBody === 'all' ||
+        config.captureBody === 'transactions'
+      ) {
+        messageCtx.body = message.value.toString();
+      }
+
+      if (message.headers) {
+        messageCtx.headers = Object.keys(message.headers).reduce(
+          (acc, name) => {
+            const value = message.headers[name];
+            if (value instanceof Buffer) {
+              acc[name] = value.toString('utf-8');
+            } else {
+              acc[name] = value;
+            }
+            return acc;
+          },
+          {},
+        );
+      }
+
+      if (message.timestamp) {
+        messageCtx.age = {
+          ms: Date.now() - Number(message.timestamp),
+        };
+      }
+
+      trans.setMessageContext(messageCtx);
+
+      let result, err;
+      try {
+        result = await origEachMessage.apply(this, arguments);
+      } catch (ex) {
+        // Save the error for use in `finally` below, but re-throw it to
+        // not impact code flow.
+        err = ex;
+        throw ex;
+      } finally {
+        trans.setOutcome(
+          err ? constants.OUTCOME_FAILURE : constants.OUTCOME_SUCCESS,
+        );
+        trans.end();
+      }
+
+      return result;
+    };
+  }
+
+  /**
+   * Returns the instrumented version of `eachBatch` which
+   * - creates a transaction each time is called
+   * - if trace context present in messages inks them to the transaction
+   *
+   * @param {ConsumerRunConfig['eachBatch']} origEachBatch
+   * @returns {ConsumerRunConfig['eachBatch']}
+   */
+  function wrapEachBatch(origEachBatch) {
+    return async function ({ batch }) {
+      if (shouldIgnoreTopic(batch.topic, config)) {
+        return origEachBatch.apply(this, arguments);
+      }
+
+      const trans = ins.startTransaction(`${NAME} RECEIVE from batch`, TYPE);
```

Comment:
```suggestion
      const trans = ins.startTransaction(`${NAME} RECEIVE from ${batch.topic}`, TYPE);
```

?   I'm not sure if there is a way to distinguish this transaction name that is receiving a batch from a transaction name for receiving a *single message*?  The spec doesn't really say: https://github.com/elastic/apm/blob/main/specs/agents/tracing-instrumentation-messaging.md#naming
Maybe a non-issue for now.

Reply:
All my tests suggested there is only one topic per batch so I?m going to apply this suggestion

Reply:
The `topic` is a property of the `batch` object, so I assume it must be just one topic -- unless the types or Kafka core docs suggest otherwise.  With a short attempt, I can't find any Kafka docs on this.